File: data_structure/search/binarysearch.go
package main

import (
	"fmt"
)

func binarySearch(a []int, v int, left, right int) bool {
	for left <= right {
		mid := (left + right) / 2
		if v == a[mid] {
			return true
		} else if v < a[mid] {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return false
}

func main() {
	a := []int{2, 3, 4, 7, 9, 12, 15, 28}
	fmt.Println(a)
	fmt.Println("Trying to find 4", binarySearch(a, 4, 0, len(a) - 1))
	fmt.Println("Trying to find 15", binarySearch(a, 15, 0, len(a) - 1))
	fmt.Println("Trying to find 28", binarySearch(a, 28, 0, len(a) - 1))
	fmt.Println("Trying to find 2", binarySearch(a, 2, 0, len(a) - 1))
	fmt.Println("Trying to find 8", binarySearch(a, 8, 0, len(a) - 1))
	fmt.Println("Trying to find 30", binarySearch(a, 30, 0, len(a) - 1))
}

/*
$ go run binarysearch.go
[2 3 4 7 9 12 15 28]
Trying to find 4 true
Trying to find 9 true
Trying to find 28 true
Trying to find 8 false
Trying to find 17 false
*/

File: data_structure/search/binarysearch.c
#include <iostream>

int find(int* a, int left, int right, int key) {
	while (left <= right) {
		int mid = (left + right) / 	2;
		if (key == a[mid]) {
			std::cout << "found " << a[mid] << ", at index: " << mid << std::endl;
			return 1;
		}	else if (key < a[mid]) {
				right = mid - 1;
		} else {
				left = mid + 1;
		}
	}
	std::cout << "not found! " << std::endl;
	return 1;
}

int main() {
	int a[] = { 1, 3, 5, 8, 10, 23, 31, 82};
	int left = 0;
	int right = sizeof(a)/sizeof(int) -1;
	std::cout << "left: " << left << ", right: " << right << std::endl;
	std::cout << "finding 0, "; find(a, left, right, 0 );
	std::cout << "finding 100, "; find(a, left, right, 100);
	std::cout << "finding 28, "; find(a, left, right, 28);
	std::cout << "finding 2, "; find(a, left, right, 2);
	std::cout << "finding 5, "; find(a, left, right, 5);
	std::cout << "finding 82, "; find(a, left, right, 82);
	std::cout << "finding 23, "; find(a, left, right, 23);
	std::cout << "finding 1, "; find(a, left, right, 1);

	return 1;
}
/*

left: 0, right: 7
finding 0, not found! 
finding 100, not found! 
finding 28, not found! 
finding 2, not found! 
finding 5, found 5, at index: 2
finding 82, found 82, at index: 7
finding 23, found 23, at index: 5
finding 1, found 1, at index: 0

*/
